Source/DafnyCompilers/CompilerRewriter.dfy

```diff
@@ -0,0 +1,2215 @@
+    // However it made me run into the following issue:
+    // BUG(https://github.com/dafny-lang/dafny/issues/2260)
+    // As ``EqValue`` appears a lot in foralls, using the `reveal` trick seemed too cumbersome
+    // to be a valid option.
+    {
+      match (v, v') {
+        case (Bool(b), Bool(b')) => b == b'
+        case (Char(c), Char(c')) => c == c'
+        case (Int(i), Int(i')) => i == i'
+        case (Real(r), Real(r')) => r == r'
+        case (BigOrdinal(o), BigOrdinal(o')) => o == o'
+        case (BitVector(width, value), BitVector(width', value')) =>
+          && width == width'
+          && value == value'
+        case (Map(m), Map(m')) =>
+          ValueTypeHeight_Children_Lem(v); // For termination
+          ValueTypeHeight_Children_Lem(v'); // For termination
+          && m.Keys == m'.Keys
+          && |m| == |m'| // We *do* need this
+          && (forall x | x in m :: EqValue(m[x], m'[x]))
+        case (Multiset(ms), Multiset(ms')) =>
+          ms == ms'
+        case (Seq(sq), Seq(sq')) =>
+          ValueTypeHeight_Children_Lem(v); // For termination
+          ValueTypeHeight_Children_Lem(v'); // For termination
+          && |sq| == |sq'|
+          && (forall i | 0 <= i < |sq| :: EqValue(sq[i], sq'[i]))
+        case (Set(st), Set(st')) =>
+          && st == st'
+        case (Closure(ctx, vars, body), Closure(ctx', vars', body')) =>
+          EqValue_Closure(v, v')
+
+        // DISCUSS: Better way to write this?  Need exhaustivity checking
+        case (Bool(b), _) => false
+        case (Char(c), _) => false
+        case (Int(i), _) => false
+        case (Real(r), _) => false
+        case (BigOrdinal(o), _) => false
+        case (BitVector(width, value), _) => false
+        case (Map(m), _) => false
+        case (Multiset(ms), _) => false
+        case (Seq(sq), _) => false
+        case (Set(st), _) => false
+        case (Closure(ctx, vars, body), _) => false
+      }
+    }
+
+    predicate {:verify true} {:opaque} EqValue_Closure(v: WV, v': WV)
+      requires v.Closure?
+      requires v'.Closure?
+      decreases ValueTypeHeight(v) + ValueTypeHeight(v'), 0
+    // Equivalence between values: closure case.
+    //
+    // See ``EqValue``.
+    //
+    // Rk.: contrary to ``EqValue``, it seems ok to make ``EqValue_Closure`` opaque.
+    {
+      var Closure(ctx, vars, body) := v;
+      var Closure(ctx', vars', body') := v';
+      && |vars| == |vars'|
+      && (
+      forall fuel:nat, argvs: seq<WV>, argvs': seq<WV> |
+        && |argvs| == |argvs'| == |vars| // no partial applications are allowed in Dafny
+        // We need the argument types to be smaller than the closure types, to prove termination.\
+        // In effect, the arguments types should be given by the closure's input types.
+        && (forall i | 0 <= i < |vars| :: ValueTypeHeight(argvs[i]) < ValueTypeHeight(v))
+        && (forall i | 0 <= i < |vars| :: ValueTypeHeight(argvs'[i]) < ValueTypeHeight(v'))
```

Comment:
Let's add a todo here to mention the fact that these should come "for free" once we have type checks?